### ISSUE TITLE
Delayed output writing "greater than or equal to" via one-stroke brief GRAOELG

### ISSUE THREAD

**Issue by jeremy-w** [openstenoproject/plover#444]
Classification: Crash/Hang/Data Loss

Reproducibility: Always (killed and relaunched, and it still happened)

## Summary

Issuing the brief for "greater than or equal to" seems to write a space, then hang for a while, then finally write. Sometimes. It's like it's hung doing the output / fake typing? My other briefs during this Anki session were shorter, but it reliably occurred with this.

A 480p screen recording demonstrating the issue should shortly be available via: http://cl.ly/433L1G3o3x42

## Steps to Reproduce
1. Launch Plover
2. Enable translation
3. Hit GRAOELG

## Expected Results
4. You should see "greater than or equal to" output immediately

## Actual Results
4. It writes a space, hangs, and eventually outputs. Sometimes it doesn't seem to manage to output. Sometimes raw keys will come through untranslated; I managed to get an "a" instead of "is" for example.

## Version

Plover version 2.5.8, but actually fbbc1f704b5cde6d1a1057ae0c66dd35783e8a02.

Installed via: git pull && sudo python launch.py

## Notes
This is a regression. I didn't have this problem till the latest git update. Before that I was on 3d28b25. Remind me to reset to that and try it again to verify.

## Configuration

OS:

ProductName:	Mac OS X
ProductVersion:	10.10.5
BuildVersion:	14F1713

**Comment by morinted** [comment]
Looks like this was caused by 03a89bb62be.

**Comment by jeremy-w** [comment]
Were you able to repro then? (And on Windows, too?)

--
Jeremy W. Sherman
http://jeremywsherman.com/

**Comment by morinted** [comment]
It's OSX-specific, I think. Happens when trying to backspace a long time, too. You can "kickstart" the output by moving the mouse.

**Comment by jeremy-w** [comment]
Ah, handy to have a workaround, thanks!
--
Jeremy W. Sherman
http://jeremywsherman.com/

**Comment by benoit-pierre** [comment]
Does this happen will all machines? What about the `plover.oslayer.keyboardcontrol` test code, does it work?

**Comment by jeremy-w** [comment]
@benoit-pierre I only have 1 NKRO keyboard, so I've no way to try across machines. I can try with the keyboardcontrol test code when I'm back at my NKRO keyboard.

**Comment by jeremy-w** [comment]
@benoit-pierre Actually, I just flipped it to use "arpeggiate" and tried it with a different keyboard. I saw the same behavior with an arpeggiated keyboard as I did with my NKRO keyboard using Plover itself.

Keyboard control gives me some wonky output and doesn't appear to do translation. It gave me stuff like:

```
> sudo python -m plover.oslayer.keyboardcontrol
Password:
Press CTRL-c to quit.
a pressed
prea released
^R    
pressed: s pressed
^R
ps released
^R      
pressed: a pressed
s pressed
^R
pesses released
da released
^R
pressed: w pressed
^R
pressed: w released
^R
prs pressed
^R
```

No clue what the `^R` were about, but they were printed literally just like that - that's not a representation of some nonprinting control character.

**Comment by benoit-pierre** [comment]
Try to give focus to a text editor during the test. You should get the 'pressed: ' string + the list of keys currently pressed down.

**Comment by morinted** [comment]
The issue is keyboard specific, I don't get the problem on the Treal.

**Comment by benoit-pierre** [comment]
OK.

**Comment by benoit-pierre** [comment]
But it's still fixed by reverting 03a89bb, no?

**Comment by jeremy-w** [comment]
Ah, it wouldn't pick up all the keys in this case anyway - I'm arpeggiating to test with a non-NKRO keyboard right now. I can check fixed by pulling master later.

**Comment by benoit-pierre** [comment]
Ah, yeah, arpeggiate is only supported by machine, not the test code.

**Comment by morinted** [comment]
It is fixed, yes. I don't mind closing it, but only if we don't aim to fix the non-blocking in the near future (again, okay with that).

**Comment by benoit-pierre** [comment]
I'd just like to know if the issue happen with the test code too, of it's another unexpected interaction with the GUI code/loop. Like for Windows, were you need to use `pythoncom.PumpWaitingMessages`, but not when the GUI is used, because it's already done by wx, as multiple callers don't work well (even in another thread). Those are important *details* with regard to wether moving the output to a dedicated engine thread is possible or not (yes on LInux, problematic on Windows, ...).

**Comment by jeremy-w** [comment]
@benoit-pierre I'm not sure what I'm looking for. When I ran the keyboard control script with my NKRO and pointed the output at a text editor, it wrote stuff like a devil possessed. Have a look: [screen recording](http://cl.ly/3c0Z0h1b1609) It did appear to register all the keypresses, though, FWIW; just output was wonky.

I tried afterward to press and hold without releasing to see if I could see just the "pressed: this+this+that", but key repeat shot that down; it started dumping just a string of `o pressed` to my terminal, and the text editor wrote gibberish.

It kept backing up past the initial writing point; sometimes I'd see a string of plusses in the midst of the output text. Basically, text output order seemed all wrong.

(This was still while sitting at fbbc1f704b5cde6d1a1057ae0c66dd35783e8a02 BTW.)

**Comment by benoit-pierre** [comment]
OK, so the output is really slow? Also, during your second test, the fact that it keeps chipping away at the text you entered manually before starting, after the first initial output, is not normal. This is what I get for comparison: https://drive.google.com/open?id=0B9Je8TDYCh1xcnlYQkN2NWRiYm8

**Comment by jeremy-w** [comment]
OK, I see with yours, it appears to backspace some once up-front, but then works in the same portion of the buffer for the rest of the test. Mine looks like it starts backing up before it's actually finished writing the current set of keys, so it backspaces too far.

**Comment by benoit-pierre** [comment]
Yeah, but since everything is taking place in the main thread, it would mean the events are not handled in the correct order? Is there a xev equivalent? Or maybe use another Python script to create an event tap and see if the order of received events is right?
